File: UE04/java/src/Labyrinth.java
import java.util.Arrays;

/**
 * Labyrinth.java by Paul Waldecker
 * 2021-11-02
 */
public class Labyrinth {
	public static String[][] maps = {{
			"############",
			"#  #     # #",
			"## # ### # #",
			"#  # # # # #",
			"## ### # # #",
			"#        # #",
			"## ####### #",
			"#          #",
			"# ######## #",
			"# #   #    #",
			"#   #   # ##",
			"######A#####"
	}, {
			"################################",
			"#                              #",
			"# ############################ #",
			"# # ###       ##  #          # #",
			"# #     ##### ### # ########## #",
			"# #   ##### #     # #      ### #",
			"# # ##### #   ###   # # ## # # #",
			"# # ### # ## ######## # ##   # #",
			"# ##### #  # #   #    #    ### #",
			"# # ### ## # # # # ####### # # #",
			"# #        # #   #     #     # #",
			"# ######## # ######### # ### # #",
			"# ####     #  # #   #  # ##### #",
			"# # #### #### # # # # ## # ### #",
			"#                      # #     #",
			"###########################A####"
	}, {
			"###########################A####",
			"#   #      ## # # ###  #     # #",
			"# ###### #### # # #### ##### # #",
			"# # ###  ## # # # #          # #",
			"# # ### ### # # # # # #### # # #",
			"# #     ### # # # # # ## # # # #",
			"# # # # ### # # # # ######## # #",
			"# # # #     #          #     # #",
			"# ### ################ # # # # #",
			"# #   #             ## # #   # #",
			"# # #### ############# # #   # #",
			"# #                    #     # #",
			"# # #################### # # # #",
			"# # #### #           ###     # #",
			"# # ## # ### ### ### ### # ### #",
			"# #    #     ##  ##  # ###   # #",
			"# ####   ###### #### # ###  ## #",
			"###########################A####"
	}, {
			"#############",
			"#           #",
			"#           #",
			"#           #",
			"###########A#"
	}};

	/**
	 * Wandelt (unveränderliche) Strings in Char-Arrays
	 * @param map  der Plan, ein String je Zeile
	 * @return char[][] des Plans
	 */
	public static char[][] fromStrings(String[] map) {
		return Arrays.stream(map).map(String::toCharArray).toArray(char[][]::new);
	}


	/**
	 * Ausgabe des Layrinths
	 * @param lab
	 */
	public static void printLabyrinth(char[][] lab) {
		Arrays.stream(lab).forEach(System.out::println);
	}

	/**
	 * Sucht, ob ein Weg aus dem Labyrinth führt
	 * @param zeile     aktuelle Position
	 * @param spalte     aktuelle Position
	 * @param lab     Labyrinth
	 * @throws InterruptedException    für die verlangsamte Ausgabe mit sleep()
	 */
	public static boolean suchen(int zeile, int spalte, char[][] lab) throws InterruptedException {
		// Checken ob A gefunden wurde
		if (lab[zeile][spalte] == 'A') {
			return true;
		}

		// Checken ob eine Wand oder eine bereits besuchte Stelle gefunden wurde
		if (lab[zeile][spalte] == '#' || lab[zeile][spalte] == '.') {
			return false;
		}

		// Die dezeite Stelle als besucht markieren
		lab[zeile][spalte] = '.';


		// Rekursiv nach dem Ausgang suchen
		boolean found = suchen(zeile - 1, spalte, lab) // Up
				|| suchen(zeile + 1, spalte, lab) // Down
				|| suchen(zeile, spalte - 1, lab) // Left
				|| suchen(zeile, spalte + 1, lab); // Right

		// aktuelle Stelle wieder freigeben
		lab[zeile][spalte] = ' ';

		return found;
	}

	/**
	 * Ermittelt die Anzahl der Wege, die aus dem Labyrinth führen
	 * @param zeile     aktuelle Position
	 * @param spalte     aktuelle Position
	 * @param lab     Labyrinth
	 * @return anzahl Anzahl der unterschiedlichen Wege, die aus dem Labyrinth führen (ein bereits besuchtes Feld darf nicht nochmals betreten werden)
	 * @throws InterruptedException    für die verlangsamte Ausgabe mit sleep()
	 */
	public static int alleSuchen(int zeile, int spalte, char[][] lab) throws InterruptedException {
		// Checken ob A gefunden wurde
		if (lab[zeile][spalte] == 'A') {
			return 1;
		}

		// Checken ob eine Wand oder eine bereits besuchte Stelle gefunden wurde
		if (lab[zeile][spalte] == '#' || lab[zeile][spalte] == '.') {
			return 0;
		}

		// Die dezeite Stelle als besucht markieren
		lab[zeile][spalte] = '.';


		// Rekursiv nach dem Ausgang suchen
		int anzahl = alleSuchen(zeile - 1, spalte, lab) // Up
				+ alleSuchen(zeile + 1, spalte, lab) // Down
				+ alleSuchen(zeile, spalte - 1, lab) // Left
				+ alleSuchen(zeile, spalte + 1, lab); // Right
		// aktuelle Stelle wieder freigeben
		lab[zeile][spalte] = ' ';

		return anzahl;
	}


	public static void main(String[] args) throws InterruptedException {
		char[][] labyrinth = fromStrings(maps[0]);
		printLabyrinth(labyrinth);
		System.out.println("Ausgang gefunden: " + (suchen(1, 1, labyrinth) ? "ja" : "nein"));
		// TODO: System.out.println("Anzahl Wege: " + suchenAlle(5, 5, labyrinth));

	}
}

File: UE04/python/Labyrinth.py
from argparse import ArgumentParser
import time

def fromStrings(map: list[str]) -> list[list[str]]:
    """
    Konvertiert eine Liste von Strings in eine Liste von Listen
    @param map: Die Liste von Strings
    @return: Die Liste von Listen
    """
    return [list(x) for x in map]

def printLabyrinth(lab: list[list[str]]) -> None:
    """
    Gibt das Labyrinth auf der Konsole aus
    @param lab: Das Labyrinth als Liste von Listen
    @return: None
    """
    for row in lab:
        print("".join(row))

def suchen(zeile: int, spalte: int, lab: list[list[str]]) -> bool:
    """
    Sucht den Ausgang im Labyrinth
    @param zeile: Die Zeile, von der aus gesucht wird
    @param spalte: Die Spalte, von der aus gesucht wird
    @param lab: Das Labyrinth als Liste von Listen
    @return: True, wenn der Ausgang gefunden wurde, sonst False
    """
    if lab[zeile][spalte] == 'A':
        return True
    if lab[zeile][spalte] in ('#', '.'):
        return False
    lab[zeile][spalte] = '.'
    found = (suchen(zeile-1, spalte, lab) or
             suchen(zeile+1, spalte, lab) or
             suchen(zeile, spalte-1, lab) or
             suchen(zeile, spalte+1, lab))
    lab[zeile][spalte] = ' '
    return found

def alleSuchen(zeile: int, spalte: int, lab: list[list[str]]) -> int:
    """
    Sucht alle möglichen Wege zum Ausgang im Labyrinth
    @param zeile: Die Zeile, von der aus gesucht wird
    @param spalte: Die Spalte, von der aus gesucht wird
    @param lab: Das Labyrinth als Liste von Listen
    @return: Die Anzahl der Wege zum Ausgang
    """
    if lab[zeile][spalte] == 'A':
        return 1
    if lab[zeile][spalte] in ('#', '.'):
        return 0
    lab[zeile][spalte] = '.'
    anzahl = (alleSuchen(zeile-1, spalte, lab) +
              alleSuchen(zeile+1, spalte, lab) +
              alleSuchen(zeile, spalte-1, lab) +
              alleSuchen(zeile, spalte+1, lab))
    lab[zeile][spalte] = ' '  # Reset the cell to its original state
    return anzahl

def main():
    # Delay fehlt
    parser = ArgumentParser(description="Calculate number of ways through a labyrinth - Paul Waldecker 5CN")
    parser.add_argument("filename", help="File containing the labyrinth to solve")
    parser.add_argument("-x", "--xstart", metavar="XSTART", type=int, help="X-coordinate to start", default=1)
    parser.add_argument("-y", "--ystart", metavar="YSTART", type=int, help="Y-coordinate to start", default=1)
    parser.add_argument("-p", "--print", action="store_true", help="Print output of every solution")
    parser.add_argument("-t", "--time", action="store_true", help="Print total calculation time (in milliseconds)")
    parser.add_argument("-d", "--delay", metavar="DELAY", type=int, help="Delay after printing a solution (in milliseconds)", default=500)
    args = parser.parse_args()

    with open(args.filename) as f:
        lab = fromStrings(f.read().splitlines())

    if args.print:
        print("Initial Labyrinth:")
        printLabyrinth(lab)
        print()

    start_zeile, start_spalte = args.ystart, args.xstart

    if args.time:
        start_time = time.time()

    if suchen(start_zeile, start_spalte, lab):
        print("Ausgang gefunden: Ja")
    else:
        print("Ausgang gefunden: Nein")

    print("Anzahl der Wege zum Ausgang:", alleSuchen(start_zeile, start_spalte, lab))

    if args.time:
        end_time = time.time()
        elapsed_time = (end_time - start_time) * 1000  # Zeit in Millisekunden
        print(f"Berechnungszeit: {elapsed_time:.2f} ms")

if __name__ == '__main__':
    main()
